libbeat/common/event_test.go

```diff
@@ -122,10 +150,149 @@ func TestConvertNestedStruct(t *testing.T) {
 				},
 			},
 		},
+		{
+			Input: MapStr{
+				"key": []interface{}{
+					TestStruct{
+						A: "hello",
+						B: 5,
+					},
+				},
+			},
+			Output: MapStr{
+				"key": []interface{}{
+					MapStr{
+						"A": "hello",
+						"B": float64(5),
+					},
+				},
+			},
+		},
 	}
 
-	for _, test := range tests {
-		assert.EqualValues(t, test.Output, ConvertToGenericEvent(test.Input))
+	for i, test := range tests {
+		assert.EqualValues(t, test.Output, ConvertToGenericEvent(test.Input), "Test case %v", i)
 	}
+}
 
+func TestNormalizeValue(t *testing.T) {
+	logp.LogInit(logp.LOG_DEBUG, "", false, true, []string{"*"})
+
+	var nilStringPtr *string
+	someString := "foo"
+
+	type mybool bool
+	type myint int32
+	type myuint uint8
+
+	var tests = []struct {
+		in  interface{}
+		out interface{}
+	}{
+		{nil, nil},
+		{&someString, someString},   // Pointers are dereferenced.
+		{nilStringPtr, nil},         // Nil pointers are dropped.
+		{NetString("test"), "test"}, // It honors the TextMarshaler contract.
+		{true, true},
+		{int8(8), int8(8)},
+		{uint8(8), uint8(8)},
+		{"hello", "hello"},
+		{map[string]interface{}{"foo": "bar"}, MapStr{"foo": "bar"}},
+
+		// Other map types are converted using marshalUnmarshal which will lose
+		// type information for arrays which become []interface{} and numbers
+		// which all become float64.
```

Comment:
We could apply the same strategy as we do in Filebeat to auto-detect ints instead of always converting to floats. See this: https://github.com/elastic/beats/blob/master/filebeat/harvester/reader/json.go#L63